File: scripts/reseed_paths.py
#!/usr/bin/env python3
from __future__ import annotations

import csv
from dataclasses import dataclass
from pathlib import Path
from typing import Mapping


ROOT = Path(__file__).resolve().parent.parent
DATA_ROOT = ROOT / "data"
RESEED_CONFIG_ROOT = DATA_ROOT / "reseed_config"
RESEED_SOURCE_ROOT = DATA_ROOT / "reseed_source"
RESEED_REPORT_ROOT = DATA_ROOT / "reseed_reports"
SQL_ROOT = ROOT / "sql"


def normalize_space(value: object) -> str:
    return " ".join(str(value or "").replace("\xa0", " ").split())


def normalize_season_name(value: object) -> str:
    return normalize_space(value).casefold()


def relative_to_root(path: Path | str) -> str:
    candidate = Path(path)
    resolved = candidate.resolve() if candidate.is_absolute() else (ROOT / candidate).resolve()
    try:
        return resolved.relative_to(ROOT).as_posix()
    except ValueError:
        return str(resolved)


@dataclass(frozen=True)
class SeasonInventoryEntry:
    season_name: str
    season_folder: str
    workbook_file: str
    ready_for_import: bool

    @property
    def season_folder_path(self) -> Path:
        return (ROOT / self.season_folder).resolve()

    @property
    def workbook_path(self) -> Path:
        return (self.season_folder_path / self.workbook_file).resolve()


def resolve_inventory_row_season_folder(row: Mapping[str, object]) -> Path:
    season_folder = normalize_space(row.get("season_folder", ""))
    if not season_folder:
        raise ValueError("Inventory row is missing season_folder.")
    return (ROOT / season_folder).resolve()


def resolve_inventory_row_workbook(row: Mapping[str, object]) -> Path:
    workbook_file = normalize_space(row.get("workbook_file", "")) or "workbook.xlsx"
    return (resolve_inventory_row_season_folder(row) / workbook_file).resolve()


def _normalize_bool(value: object) -> bool:
    return normalize_space(value).upper() in {"1", "TRUE", "YES", "Y"}


def read_season_inventory(config_root: Path = RESEED_CONFIG_ROOT) -> list[SeasonInventoryEntry]:
    inventory_path = config_root / "season_inventory.csv"
    rows: list[SeasonInventoryEntry] = []
    if not inventory_path.exists():
        return rows

    with inventory_path.open(newline="", encoding="utf-8-sig") as handle:
        reader = csv.DictReader(handle)
        for row in reader:
            season_name = normalize_space(row.get("season_name", ""))
            if not season_name:
                continue

            rows.append(
                SeasonInventoryEntry(
                    season_name=season_name,
                    season_folder=normalize_space(row.get("season_folder", "")),
                    workbook_file=normalize_space(row.get("workbook_file", "")) or "workbook.xlsx",
                    ready_for_import=_normalize_bool(row.get("ready_for_import")),
                )
            )

    return rows


def resolve_season_inventory_entry(
    season_name: str,
    config_root: Path = RESEED_CONFIG_ROOT,
) -> SeasonInventoryEntry:
    normalized_target = normalize_season_name(season_name)
    entries = read_season_inventory(config_root)
    for entry in entries:
        if normalize_season_name(entry.season_name) == normalized_target:
            return entry

    known_seasons = ", ".join(entry.season_name for entry in entries) or "(none)"
    raise FileNotFoundError(
        f"Season {season_name!r} was not found in {relative_to_root(config_root / 'season_inventory.csv')}. "
        f"Known seasons: {known_seasons}"
    )


def resolve_season_workbook(
    season_name: str,
    config_root: Path = RESEED_CONFIG_ROOT,
) -> Path:
    return resolve_season_inventory_entry(season_name, config_root).workbook_path

File: scripts/reseed_paths.js
const fs = require("node:fs");
const path = require("node:path");

const ROOT = path.resolve(__dirname, "..");
const DATA_ROOT = path.join(ROOT, "data");
const RESEED_CONFIG_ROOT = path.join(DATA_ROOT, "reseed_config");
const RESEED_SOURCE_ROOT = path.join(DATA_ROOT, "reseed_source");
const RESEED_REPORT_ROOT = path.join(DATA_ROOT, "reseed_reports");
const SEASON_INVENTORY_PATH = path.join(RESEED_CONFIG_ROOT, "season_inventory.csv");

function normalizeSpace(value) {
  return String(value || "")
    .replace(/\u00a0/g, " ")
    .trim()
    .split(/\s+/)
    .filter(Boolean)
    .join(" ");
}

function normalizeSeasonName(value) {
  return normalizeSpace(value).toUpperCase();
}

function relativeToRoot(targetPath) {
  const resolved = path.resolve(targetPath);
  const relative = path.relative(ROOT, resolved);
  return relative && !relative.startsWith("..") && !path.isAbsolute(relative) ? relative : resolved;
}

function parseCsvLine(line) {
  const cells = [];
  let current = "";
  let inQuotes = false;

  for (let index = 0; index < line.length; index += 1) {
    const character = line[index];
    const nextCharacter = line[index + 1];

    if (character === '"') {
      if (inQuotes && nextCharacter === '"') {
        current += '"';
        index += 1;
        continue;
      }
      inQuotes = !inQuotes;
      continue;
    }

    if (character === "," && !inQuotes) {
      cells.push(current);
      current = "";
      continue;
    }

    current += character;
  }

  cells.push(current);
  return cells;
}

function readSeasonInventory(configRoot = RESEED_CONFIG_ROOT) {
  const inventoryPath = path.join(configRoot, "season_inventory.csv");
  if (!fs.existsSync(inventoryPath)) {
    return [];
  }

  const lines = fs
    .readFileSync(inventoryPath, "utf8")
    .split(/\r?\n/)
    .filter((line, index) => index === 0 || line.trim());

  if (!lines.length) {
    return [];
  }

  const header = parseCsvLine(lines[0]);
  return lines.slice(1).map((line) => {
    const values = parseCsvLine(line);
    const row = Object.fromEntries(header.map((column, index) => [column, values[index] || ""]));
    return {
      season_name: normalizeSpace(row.season_name),
      season_folder: normalizeSpace(row.season_folder),
      workbook_file: normalizeSpace(row.workbook_file) || "workbook.xlsx",
      ready_for_import: ["1", "TRUE", "YES", "Y"].includes(normalizeSpace(row.ready_for_import).toUpperCase()),
    };
  });
}

function resolveSeasonInventoryEntry(seasonName, configRoot = RESEED_CONFIG_ROOT) {
  const normalizedTarget = normalizeSeasonName(seasonName);
  const rows = readSeasonInventory(configRoot);
  const match = rows.find((row) => normalizeSeasonName(row.season_name) === normalizedTarget);
  if (match) {
    return match;
  }

  const knownSeasons = rows.map((row) => row.season_name).join(", ") || "(none)";
  throw new Error(
    `Season ${JSON.stringify(seasonName)} was not found in ${relativeToRoot(
      path.join(configRoot, "season_inventory.csv")
    )}. Known seasons: ${knownSeasons}`
  );
}

function resolveSeasonWorkbookPath(seasonName, configRoot = RESEED_CONFIG_ROOT) {
  const entry = resolveSeasonInventoryEntry(seasonName, configRoot);
  return path.resolve(ROOT, entry.season_folder, entry.workbook_file);
}

module.exports = {
  ROOT,
  DATA_ROOT,
  RESEED_CONFIG_ROOT,
  RESEED_SOURCE_ROOT,
  RESEED_REPORT_ROOT,
  SEASON_INVENTORY_PATH,
  normalizeSpace,
  relativeToRoot,
  readSeasonInventory,
  resolveSeasonInventoryEntry,
  resolveSeasonWorkbookPath,
};
